rgrow/src/models/sdc1d.rs

```diff
@@ -0,0 +1,805 @@
+        // conc = stoic * u0 * exp(-Gmc + alpha) and u0 = 1M, but we really should just have
+        // a way to specify concentrations directly.
+        let strand_concentration = pc
+            .tile_stoics
+            .mapv(|x| x * (-tileset.gmc.unwrap_or(16.0) + alpha).exp());
+
+        let mut sys = SDC {
+            strand_names: pc.tile_names,
+            glue_names: pc.glue_names,
+            glue_links,
+            colors: pc.tile_colors,
+            glues: pc.tile_edges,
+            anchor_tiles: Vec::new(),
+            scaffold,
+            strand_concentration,
+            kf: tileset.kf.unwrap_or(1.0e6),
+            delta_g_matrix: todo!(),
+            entropy_matrix: todo!(),
+            temperature: todo!(),
+            friends_btm: HashMap::new(),
+            strand_energy_bonds: energy_bonds,
+            scaffold_energy_bonds: todo!(),
+        };
+
+        // This will generate the friends hashamp, as well as the glues, and the energy bonds
+        sys.update_system();
+
+        Ok(sys)
+    }
+}
+
+// Here is potentially another way to process this, though not done.  Feel free to delete or modify.
+
+use std::hash::Hash;
+
+use bimap::BiHashMap;
+
+#[cfg(python)]
+use pyo3::prelude::*;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(python, derive(FromPyObject))]
+pub enum RefOrPair {
+    Ref(String),
+    Pair(String, String),
+}
+
+impl From<String> for RefOrPair {
+    fn from(r: String) -> Self {
+        RefOrPair::Ref(r)
+    }
+}
+
+impl From<(String, String)> for RefOrPair {
+    fn from(p: (String, String)) -> Self {
+        RefOrPair::Pair(p.0, p.1)
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(python, derive(FromPyObject))]
+pub enum SingleOrMultiScaffold {
+    Single(Vec<String>),
+    Multi(Vec<Vec<String>>),
+}
+
+impl From<Vec<String>> for SingleOrMultiScaffold {
+    fn from(v: Vec<String>) -> Self {
+        SingleOrMultiScaffold::Single(v)
+    }
+}
+
+impl From<Vec<Vec<String>>> for SingleOrMultiScaffold {
+    fn from(v: Vec<Vec<String>>) -> Self {
+        SingleOrMultiScaffold::Multi(v)
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(python, derive(FromPyObject))]
+pub struct SDCParams {
+    pub tile_glues: Vec<Vec<Option<String>>>,
+    pub tile_concentration: Vec<f64>,
+    pub tile_names: Vec<Option<String>>,
+    pub tile_colors: Vec<Option<String>>,
+    pub scaffold: SingleOrMultiScaffold,
+    // Pair with delta G at 37 degrees C and delta S
+    pub glue_dg_s: HashMap<RefOrPair, (f64, f64)>,
+    pub k_f: f64,
+    pub k_n: f64,
+    pub k_c: f64,
+    pub temperature: f64,
+}
+
+// This is not a smart way of doing this, but it works for now!
+#[cfg(python)]
+#[pymethods]
+impl SystemEnum {
+    fn new_sdc(params: SDCParams) -> SystemEnum {
+        SystemEnum::SDC(SDC::from_params(params))
+    }
+}
+
+/// Return the orignial, and its inverse
+fn self_and_inverse(value: &String) -> (String, String) {
```

Comment:
The logic for this function (may be wrong) is that the inverse of the inverse is the original. 

So given h*, its inverse is just h, not h**. This makes it so that having more than one * is pointless.

I assume this work since h* and h have a perfect bond, and h* and (h*)* have a perfect bond, then h** = h.

Reply:
Yes, that works.  The one potential problem is if a user gives an input of "g**" somewhere.  But **they shouldn't do that.**